File: 95-unique-binary-search-trees-ii/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        if n == 0: return []
        
        def build(left, right) -> List[Optional[TreeNode]]:
            res = []
            if left > right:
                res.append(None)
                return res
            
            for i in range(left, right+1):
                left_nodes = build(left, i-1)
                rigth_nodes = build(i+1, right)
                for left_node in left_nodes:
                    for right_node in rigth_nodes:
                        root = TreeNode(i)
                        root.left = left_node
                        root.right = right_node
                        res.append(root)
            return res
        return build(1, n)

File: 95-unique-binary-search-trees-ii/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if (n == 0) return vector<TreeNode*>();
        return build(1, n);
    }

    vector<TreeNode*> build(int low, int high) {
        vector<TreeNode*> res;
        if (low > high) {
            res.push_back(nullptr);
            return res;
        }

        for (int i = low; i <= high; i ++) {
            vector<TreeNode*> leftNodes = build(low, i-1);
            vector<TreeNode*> rightNodes = build(i+1, high);
            for (TreeNode* leftNode: leftNodes) {
                for (TreeNode* rightNode: rightNodes) {
                    TreeNode* root = new TreeNode(i);
                    root->left = leftNode;
                    root->right = rightNode;
                    res.push_back(root);
                }
            }
        }
        return res;
    }

};